### PR TITLE
fix(ActiveSelection): 🚀 multiple nested selection 🗂️

### LINKED ISSUE #6776: Create ActiveSelection with objects in group
I want to create a multiple selection from objects in group.
After create selection with ActiveSelection, bound box of the ActiveSelection is wrong. 
How to select multiple object within group or nested group?

check it here.
https://jsfiddle.net/zhongjian2008/49wdsg2q/

### PR DESCRIPTION
**Pull request by ShaMan123**
# Multiple Nested Selection
This PR finalizes (so it seems) the group rewrite!

## Gist

- refactored `_chooseObjectsToRender`
- made canvas hold a constant ref to `ActiveSelection` instead of creating new ones all time (important for devs wanting to attach events to it for customization, see #7878)
- introduced `Group#invalidate` - to handle children/ActiveSelection state changes that should propagate to group + an option for layout customization (`progress` - very experimental and unstable at the moment, remains optional)
- invalidate owning group of object under ActiveSelection 

The `progess` layout trigger is designed to enable updating group's size dynamically but it introduces unwanted changes and craziness. e.g. when object is rotating. We should consider if and how to support it. I think it should remain as is, the dev can override `getLayoutStrategyResult` and do what they want.

## TODO
- [x] figure out what to do with exporters (toDataUrl, toCanvas) on group while some objects are selected and `canvas.preserveObjectStacking = false` and therefore not rendered by group - **DONE** 79db7d5b
- [ ] `test/visual/golden/group-layout/selected_object.png` - seems the selected rect is doubled in opacity...
- [ ] rethink ActiveSelection caching - why not?

## Dev

Video was heavy so visuals are poor but perf is amazing
ezgif com-gif-maker (1)

closes #6776

[Edit fabric_app (forked)](https://codesandbox.io/s/fabric-app-forked-i29woj?fontsize=14&hidenavigation=1&module=%2Fsrc%2FManyLayouts.tsx&theme=dark)

<details> <summary>test code to paste in the kitchensink demo</summary>

```js

// clear canvas
canvas.clear();
 canvas.preserveObjectStacking = false;
    const circle = new fabric.Circle({ left: 100, top: 50, radius: 50 });
    const text = new fabric.Text('', { evented: false });
    const itext = new fabric.IText('Edit me\nfit-content layout', { left: 100, top: 150 });
    const g = new fabric.Group([
      new fabric.Rect({
        top: 200,
        width: 50,
        height: 50,
        fill: 'red',
        opacity: 0.3
      }),
      circle,
      itext
    ], {
      backgroundColor: 'blue',
      subTargetCheck: true,
      interactive: true
      //objectCaching: false
    });
    canvas.add(
      text,
      g
    );
    g.clone().then(clone => {
      clone.item(2).set({ text: 'Edit me\nclip path layout' });
      clone.set({
        backgroundColor: 'magenta',
        clipPath: new fabric.Circle({
          radius: 110,
          group: clone
          //left: -50, top: -50,
          // scaleX:0.6
        }),
        layout: 'clip-path'
      });
     // canvas.add(clone);
    });
    g.clone().then(clone => {
      clone.item(2).set({ text: 'Edit me\nclip path layout' });
      clone.set({
        backgroundColor: 'yellow',
        clipPath: new fabric.Circle({
          radius: 110,
          //left: -50, top: -50,
          originX: 'center', originY: 'center',
          group: clone
        }),
        layout: 'clip-path'
      });
      canvas.add(clone);
      setInterval(() => {
        clone._set('dirty', true)
      }, 0);
    });
    g.clone().then(clone => {
      clone.item(2).set({ text: 'Edit me\nabsolute positioned\nclip path layout' });
      clone.set({
        backgroundColor: '#0dcaf0',
        clipPath: new fabric.Circle({
          radius: 110, originX: 'center', originY: 'center',
          absolutePositioned: true, left: 50, top: 150, skewX:20
        }),
        layout: 'clip-path',

      });
      canvas.insertAt(clone, 0);
    });
    canvas.on('after:render', () => {
      text.set('text', `circle is on screen? ${circle.isOnScreen()}`);
      text.dirty && canvas.requestRenderAll();
    });
```
</details>

### CODE REVIEW

**Review by ShaMan123 — COMMENTED** [review]
updated from master

File: src/shapes/active_selection.class.js [hunk 1]
```diff
@@ -52,13 +52,49 @@
...
+          target: this,
+          key: key,
+          value: value,
+          prevValue: prev
+        };
+        var invalidatedGroups = [];
+        this.forEachObject(function (object) {
+          var group = object.__owningGroup;
+          if (group && invalidatedGroups.indexOf(group) === -1) {
+            group.invalidate(invalidationContext);
```
Comment (ShaMan123) [inline src/shapes/active_selection.class.js:75]: maybe I want to add a check for objects that are group de-facto but aren't really groups:
```suggestion
            group.invalidate && group.invalidate(invalidationContext);
```

**Review by ShaMan123 — COMMENTED** [review]
`preservevObjectStacking` plays a role in this PR in group rendering logic.
We can work around it, we need to think how though

**Comment by ShaMan123** [comment]
This PR will wait until caching is decided #7874

File: src/shapes/active_selection.class.js [hunk 2]
```diff
@@ -52,13 +52,49 @@
       this.setCoords();
     },
 
+    /**
+     * @private
+     * @param {string} key
+     * @param {*} value
+     */
+    _set: function (key, value) {
```
Comment (ShaMan123) [inline src/shapes/active_selection.class.js:60]: this is the only part of active selection that hasn't been ported to #8665

**Review by ShaMan123 — COMMENTED** [review]
files to port to #8665 :
- [ ] `src/shapes/active_selection.class.js`
- [ ] `src/shapes/group.class.js`
- [ ] `src/shapes/object.class.js`
- [ ] `test/unit/activeselection.js`
- [ ] `test/unit/util.js`
- [ ] `test/visual/group_layout.js` - not sure what the test is

**Comment by ShaMan123** [comment]
what remains to port from this PR is caching logic and maybe a bit of rendering

File: src/shapes/group.class.js [hunk 3]
```diff
@@ -257,6 +257,11 @@
         object.group.remove(object);
       }
       this._enterGroup(object, removeParentTransform);
+      var activeObject = this.canvas && this.canvas.getActiveObject && this.canvas.getActiveObject();
+      // if we are adding the activeObject in a group
+      if (activeObject && (activeObject === object || object.isDescendantOf(activeObject))) {
+        this._activeObjects.push(object);
+      }
```
Comment (ShaMan123) [inline src/shapes/group.class.js:264]: this part....\

**Comment by asturur** [comment]
capturing this test case here since the functionality is not working with the current layout manager and adding an issue to describe it. The code here is not reconciliable

**Comment by ShaMan123** [comment]
Reviewing the code seems we can test to see if it works because I don't see too much here that isn't in the codebase

**Pull request closed without merging**